File: feral-nodes/ts-node-sdk/src/schemas.ts
/*
 * Zod mirrors of the HUP v1 wire schemas from ../../HUP_SPEC.md §5.
 * These validate outbound frames before send so daemons are conformant by construction.
 */

import { z } from "zod";

export const HUP_VERSION = "1.2.0";

// Per-frame decoded-size caps from HUP_SPEC.md §5.4.1 / §5.4.2.
export const AUDIO_FRAME_MAX_BYTES = 64 * 1024;
export const VIDEO_FRAME_MAX_BYTES = 512 * 1024;

function decodedBase64Size(b64: string): number {
  // Cheap tight upper bound: bytes = (len * 3) / 4 - padding count.
  const len = b64.length;
  const padding = (b64.endsWith("==") ? 2 : b64.endsWith("=") ? 1 : 0);
  return Math.floor((len * 3) / 4) - padding;
}

export const NodeType = z.enum([
  "desktop",
  "server",
  "rpi",
  "robot",
  "glasses",
  "phone",
  "actuator",
  "sensor",
  "wearable",
  "camera",
  "vehicle",
  "appliance",
]);
export type NodeType = z.infer<typeof NodeType>;

export const NodeRegisterPayload = z.object({
  node_id: z.string().regex(/^[A-Za-z0-9._:-]{1,128}$/),
  node_type: NodeType.default("sensor"),
  name: z.string().default(""),
  manufacturer: z.string().default(""),
  model: z.string().default(""),
  firmware_version: z.string().default(""),
  platform: z.string().default(""),
  os: z.string().default(""),
  capabilities: z.array(z.string()).default([]),
  sensors: z.array(z.string()).default([]),
  actuators: z.array(z.string()).default([]),
  location: z.string().default(""),
  tags: z.array(z.string()).default([]),
});
export type NodeRegisterPayload = z.infer<typeof NodeRegisterPayload>;

export const NodeAckPayload = z.object({
  node_id: z.string(),
  session_token: z.string(),
  heartbeat_ms: z.number().int().default(10_000),
  server_time: z.number().default(() => Date.now() / 1000),
  granted_capabilities: z.array(z.string()).default([]),
  denied_capabilities: z.array(z.string()).default([]),
});
export type NodeAckPayload = z.infer<typeof NodeAckPayload>;

export const NodeHeartbeatPayload = z.object({
  ts: z.number().default(() => Date.now() / 1000),
  battery_pct: z.number().int().min(0).max(100).nullable().optional(),
  rssi: z.number().int().nullable().optional(),
});
export type NodeHeartbeatPayload = z.infer<typeof NodeHeartbeatPayload>;

export const DeviceEventPayload = z.object({
  node_id: z.string(),
  event_type: z.string(),
  data: z.record(z.any()).default({}),
  ts: z.number().default(() => Date.now() / 1000),
});
export type DeviceEventPayload = z.infer<typeof DeviceEventPayload>;

// HUP v1.1 audio_frame payload — per HUP_SPEC.md §5.4.1.
export const AudioFramePayload = z.object({
  event_type: z.literal("audio_frame").default("audio_frame"),
  codec: z.enum(["opus", "pcm16"]),
  sample_rate: z.number().int().min(8000).max(96000),
  channels: z.number().int().min(1).max(2),
  frame_ms: z.number().int().min(1).max(120).default(20),
  sequence: z.number().int().min(0),
  data_b64: z.string().refine(
    (v) => decodedBase64Size(v) <= AUDIO_FRAME_MAX_BYTES,
    { message: `audio_frame data_b64 exceeds ${AUDIO_FRAME_MAX_BYTES} bytes decoded` },
  ),
});
export type AudioFramePayload = z.infer<typeof AudioFramePayload>;

// HUP v1.1 video_frame payload — per HUP_SPEC.md §5.4.2.
export const VideoFramePayload = z.object({
  event_type: z.literal("video_frame").default("video_frame"),
  codec: z.enum(["jpeg", "h264"]),
  width: z.number().int().min(1).max(8192),
  height: z.number().int().min(1).max(8192),
  sequence: z.number().int().min(0),
  keyframe: z.boolean().default(true),
  data_b64: z.string().refine(
    (v) => decodedBase64Size(v) <= VIDEO_FRAME_MAX_BYTES,
    { message: `video_frame data_b64 exceeds ${VIDEO_FRAME_MAX_BYTES} bytes decoded` },
  ),
});
export type VideoFramePayload = z.infer<typeof VideoFramePayload>;

export const HUPActionRequestPayload = z.object({
  action_id: z.string().min(1).max(64),
  name: z.string().min(1).max(64),
  params: z.record(z.any()).default({}),
  timeout_ms: z.number().int().min(1).max(120_000).default(5000),
  requires_confirmation: z.boolean().default(false),
});
export type HUPActionRequestPayload = z.infer<typeof HUPActionRequestPayload>;

export const HUPActionResponsePayload = z.object({
  action_id: z.string(),
  success: z.boolean(),
  result: z.record(z.any()).default({}),
  error: z.string().nullable().optional(),
  duration_ms: z.number().int().min(0).default(0),
});
export type HUPActionResponsePayload = z.infer<typeof HUPActionResponsePayload>;

export const NodeByePayload = z.object({
  reason: z.string().default("shutdown"),
  restart_in_s: z.number().int().default(0),
});
export type NodeByePayload = z.infer<typeof NodeByePayload>;

export const ErrorPayload = z.object({
  code: z.number().int(),
  name: z.string(),
  message: z.string(),
  recoverable: z.boolean().default(true),
  ref_action_id: z.string().nullable().optional(),
});
export type ErrorPayload = z.infer<typeof ErrorPayload>;

export const HUPMessageType = z.enum([
  "node_register",
  "node_ack",
  "node_heartbeat",
  "device_event",
  "hup_action_request",
  "hup_action_response",
  "node_bye",
  "error",
]);
export type HUPMessageType = z.infer<typeof HUPMessageType>;

const SCHEMAS: Record<string, z.ZodTypeAny> = {
  node_register: NodeRegisterPayload,
  node_ack: NodeAckPayload,
  node_heartbeat: NodeHeartbeatPayload,
  device_event: DeviceEventPayload,
  hup_action_request: HUPActionRequestPayload,
  hup_action_response: HUPActionResponsePayload,
  node_bye: NodeByePayload,
  error: ErrorPayload,
};

export interface HUPFrame {
  hup_version: string;
  type: string;
  ts: number;
  payload: Record<string, unknown>;
}

export function buildFrame(
  type: string,
  payload: Record<string, unknown>,
): HUPFrame {
  const schema = SCHEMAS[type];
  if (!schema) {
    throw new Error(`unknown HUP message type: ${type}`);
  }
  const parsed = schema.parse(payload) as Record<string, unknown>;
  return {
    hup_version: HUP_VERSION,
    type,
    ts: Date.now() / 1000,
    payload: parsed,
  };
}

File: feral-nodes/python-node-sdk/src/feral_node_sdk/schemas.py
"""Pydantic mirrors of the HUP v1 wire schemas in `HUP_SPEC.md` §5.

These models are the canonical *runtime* validators for the Python SDK; the
Markdown spec is the canonical *normative* source. Keep them in lockstep.
"""

from __future__ import annotations

import time as _time
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, field_validator

HUP_VERSION = "1.1.0"
# Per-frame decoded-size caps from HUP_SPEC.md §5.4.1 / §5.4.2.
AUDIO_FRAME_MAX_BYTES = 64 * 1024
VIDEO_FRAME_MAX_BYTES = 512 * 1024

NodeType = Literal[
    "desktop",
    "server",
    "rpi",
    "robot",
    "glasses",
    "phone",
    "actuator",
    "sensor",
    "wearable",
    "camera",
    "vehicle",
    "appliance",
]


class HUPFrame(BaseModel):
    """Outer envelope shared by every HUP message."""

    hup_version: str = HUP_VERSION
    type: str
    ts: float = Field(default_factory=_time.time)
    payload: dict[str, Any] = Field(default_factory=dict)


class NodeRegisterPayload(BaseModel):
    node_id: str = Field(..., pattern=r"^[A-Za-z0-9._:-]{1,128}$")
    node_type: NodeType = "sensor"
    name: str = ""
    manufacturer: str = ""
    model: str = ""
    firmware_version: str = ""
    platform: str = ""
    os: str = ""
    capabilities: list[str] = Field(default_factory=list)
    sensors: list[str] = Field(default_factory=list)
    actuators: list[str] = Field(default_factory=list)
    location: str = ""
    tags: list[str] = Field(default_factory=list)

    @field_validator("capabilities", "sensors", "actuators", "tags", mode="before")
    @classmethod
    def _to_str_list(cls, v: Any) -> list[str]:
        if v is None:
            return []
        return [str(x) for x in v]


class NodeAckPayload(BaseModel):
    node_id: str
    session_token: str
    heartbeat_ms: int = 10000
    server_time: float = Field(default_factory=_time.time)
    granted_capabilities: list[str] = Field(default_factory=list)
    denied_capabilities: list[str] = Field(default_factory=list)


class NodeHeartbeatPayload(BaseModel):
    ts: float = Field(default_factory=_time.time)
    battery_pct: Optional[int] = Field(default=None, ge=0, le=100)
    rssi: Optional[int] = None


class DeviceEventPayload(BaseModel):
    node_id: str
    event_type: str
    data: dict[str, Any] = Field(default_factory=dict)
    ts: float = Field(default_factory=_time.time)


class AudioFramePayload(BaseModel):
    """HUP v1.1 `audio_frame` payload (per HUP_SPEC.md §5.4.1)."""

    event_type: Literal["audio_frame"] = "audio_frame"
    codec: Literal["opus", "pcm16"]
    sample_rate: int = Field(..., ge=8000, le=96000)
    channels: int = Field(..., ge=1, le=2)
    frame_ms: int = Field(default=20, ge=1, le=120)
    sequence: int = Field(..., ge=0)
    data_b64: str

    @field_validator("data_b64")
    @classmethod
    def _data_b64_decoded_size(cls, v: str) -> str:
        import base64
        try:
            decoded = base64.b64decode(v, validate=False)
        except Exception as exc:
            raise ValueError(f"data_b64 is not valid base64: {exc}") from exc
        if len(decoded) > AUDIO_FRAME_MAX_BYTES:
            raise ValueError(
                f"audio_frame data_b64 decoded to {len(decoded)} bytes; cap is {AUDIO_FRAME_MAX_BYTES}"
            )
        return v


class VideoFramePayload(BaseModel):
    """HUP v1.1 `video_frame` payload (per HUP_SPEC.md §5.4.2)."""

    event_type: Literal["video_frame"] = "video_frame"
    codec: Literal["jpeg", "h264"]
    width: int = Field(..., ge=1, le=8192)
    height: int = Field(..., ge=1, le=8192)
    sequence: int = Field(..., ge=0)
    keyframe: bool = True
    data_b64: str

    @field_validator("data_b64")
    @classmethod
    def _data_b64_decoded_size(cls, v: str) -> str:
        import base64
        try:
            decoded = base64.b64decode(v, validate=False)
        except Exception as exc:
            raise ValueError(f"data_b64 is not valid base64: {exc}") from exc
        if len(decoded) > VIDEO_FRAME_MAX_BYTES:
            raise ValueError(
                f"video_frame data_b64 decoded to {len(decoded)} bytes; cap is {VIDEO_FRAME_MAX_BYTES}"
            )
        return v


class HUPActionRequestPayload(BaseModel):
    action_id: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=64)
    params: dict[str, Any] = Field(default_factory=dict)
    timeout_ms: int = Field(default=5000, ge=1, le=120_000)
    requires_confirmation: bool = False


class HUPActionResponsePayload(BaseModel):
    action_id: str
    success: bool
    result: dict[str, Any] = Field(default_factory=dict)
    error: Optional[str] = None
    duration_ms: int = Field(default=0, ge=0)


class NodeByePayload(BaseModel):
    reason: str = "shutdown"
    restart_in_s: int = 0


class ErrorPayload(BaseModel):
    code: int
    name: str
    message: str
    recoverable: bool = True
    ref_action_id: Optional[str] = None


MESSAGE_TYPES: dict[str, type[BaseModel]] = {
    "node_register": NodeRegisterPayload,
    "node_ack": NodeAckPayload,
    "node_heartbeat": NodeHeartbeatPayload,
    "device_event": DeviceEventPayload,
    "hup_action_request": HUPActionRequestPayload,
    "hup_action_response": HUPActionResponsePayload,
    "node_bye": NodeByePayload,
    "error": ErrorPayload,
}


def build_frame(type_: str, payload: BaseModel | dict[str, Any]) -> dict[str, Any]:
    """Serialize a validated payload into a wire-ready HUP envelope."""
    model_cls = MESSAGE_TYPES.get(type_)
    if model_cls is None:
        raise ValueError(f"unknown HUP message type: {type_!r}")
    if isinstance(payload, BaseModel):
        validated = payload
    else:
        validated = model_cls(**payload)
    return {
        "hup_version": HUP_VERSION,
        "type": type_,
        "ts": _time.time(),
        "payload": validated.model_dump(exclude_none=False),
    }
